docs/proposals/pod-security-context.md

```diff
@@ -0,0 +1,154 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+A proposal for refactoring `SecurityContext` to have pod-level and container-level facets in order
+to correctly model pod- and container-level security concerns.
+
+## Motivation
+
+Currently, containers have a `SecurityContext` attribute which contains information about the
+security settings the container uses.  A container-level context lacks the ability, however, to
+express notions that are really pod-level.  A pod-level security context will:
+
+1.  Allow users to cleanly express security settings that are applicable to the entire pod
+2.  Allow users to express default settings for the container-level context and override only the
+    settings that need to be differentiated from the pod security context in containers where
+    required
+
+Goals of this design:
+
+1.  Describe the use-cases for which a pod-level security context is necessary
+2.  Describe the model changes and associated implementation changes for the refactor
+
+## Use Cases
+
+1.  As a user, I want to be able to specify a default security context for all containers in a pod
+2.  As a user, I want to be able to override applicable pieces of the pod security context for
+    individual containers as needed
+
+### Use Case: Default security context for all containers
+
+By default, all containers in a pod should share the same security settings, and those settings
+should only have to be specified once.  If users are forced to specify the same settings for each
+container, it requires more complex validations and allows for transcription errors between
+containers
+
+Additionally, use-cases for sharing resources such as volumes are much easier to implement if a user
+can describe a pod-level security context.
+
+### Use Case: Override pod security context for container
+
+Some use-cases require the containers in a pod to run with different security settings.  As an
+example, a user may want to have a pod with two containers, one of which runs as root with the
+privileged setting, and one that runs as a non-root UID.  To support use-cases like this, it should
+be possible to override appropriate (ie, not intrinsically pod-level) security settings for
+individual containers.
+
+## Proposed Design
+
+The `SecurityContext` type should be replaced by two new, similar types: `ContainerSecurityContext`
+and `PodSecurityContext`:
+
+### ContainerSecurityContext
+
+The `ContainerSecurityContext` type models the security settings that specified for a container:
+
+```go
+type ContainerSecurityContext struct {
+    Capabilities              *Capabilities
+    Privileged                *bool
+    SELinuxOptions            *SELinuxOptions
+    RunAsUser                 *int64
+    RunAsNonRoot              bool
```

Comment:
Good question -- they are two different facets:

1.   `RunAsUser` is the specific UID that is requested
2.   `RunAsNonRoot` is to ensure that when the uid is delegated to what's declared in the image metadata, that the image doesn't want to run as root.

If I had included the comments on those fields it would probably be clearer.